share/p2p/discovery/discovery.go

```diff
@@ -202,21 +204,32 @@ func (d *Discovery) Advertise(ctx context.Context) {
 // It initiates peer discovery upon request and restarts the process until the soft limit is
 // reached.
 func (d *Discovery) discoveryLoop(ctx context.Context) {
-	t := time.NewTicker(discoveryRetryTimeout)
-	defer t.Stop()
+	backoffDuration := backoffInitialDuration
+	backoffTicker := time.NewTicker(backoffDuration)
+	defer backoffTicker.Stop()
 
 	warnTicker := time.NewTicker(logInterval)
 	defer warnTicker.Stop()
 
+	// Function to initiate discovery and restart the timer
+	runDiscovery := func() {
+		if !d.discover(ctx) {
+			// rerun discovery if the number of peers hasn't reached the limit
+			return
```

Comment:
This `return` stop execution only of anonymous func, when you would want to return from outer loop